src/writing-your-own-application/no-std-applications/understanding-esp-template.md

```diff
@@ -0,0 +1,198 @@
+
+```text
+│   .gitignore
+│   Cargo.toml
+│   LICENSE-APACHE
+│   LICENSE-MIT
+│   rust-toolchain.toml
+│
+├───.cargo
+│       config.toml
+│
+├───.vscode
+│       settings.json
+│
+└───src
+        main.rs
+```
+
+
+Before going further let's see what these files are for.
+
+- [.gitignore]
+    - tells `git` which folders and files to ignore
+- [Cargo.toml]
+    - the usual Cargo manifest declaring some meta-data and dependencies of the project
+- LICENSE-APACHE, LICENSE_MIT
+    - those are the most common licenses used in the Rust ecosystem
+    - if you want to apply a different license you can delete these files and change the license in `Cargo.toml`
+- [rust-toolchain.toml]
+    - defines which Rust toolchain to use
+    - depending on your target this will use `nightly` or `esp`
+- [.cargo/config.toml]
+    - the Cargo configuration
+    - this defines a few options to correctly build the project
+    - also contains `runner = "espflash --monitor"` - this means you can just use `cargo run` to flash and monitor your code
+- .vscode/settings.json
+    - settings for Visual Studio Code - if you are not using VSCode you can delete the whole folder
+- src/main.rs
+    - the main source file of the newly created project
+    - we will examine its content in the next section
+
+## `main.rs`
+
+```rust,ignore
+#![no_std]
+#![no_main]
+
+use esp32c3_hal::{clock::ClockControl, pac::Peripherals, prelude::*, timer::TimerGroup, Rtc};
+use esp_backtrace as _;
+
+#[riscv_rt::entry]
+fn main() -> ! {
+    let peripherals = Peripherals::take().unwrap();
+    let system = peripherals.SYSTEM.split();
+    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
+
+    // Disable the RTC and TIMG watchdog timers
+    let mut rtc = Rtc::new(peripherals.RTC_CNTL);
+    let timer_group0 = TimerGroup::new(peripherals.TIMG0, &clocks);
+    let mut wdt0 = timer_group0.wdt;
+    let timer_group1 = TimerGroup::new(peripherals.TIMG1, &clocks);
+    let mut wdt1 = timer_group1.wdt;
+
+    rtc.swd.disable();
+    rtc.rwdt.disable();
+    wdt0.disable();
+    wdt1.disable();
+
+    loop {}
+}
+```
+
+That is quite a lot of code. Let's see what it is good for.
+
+- `#![no_std]`
+    - this tells the Rust compiler that this code doesn't use `libstd`
+- `#![no_main]`
+    - The `no_main` attribute says that this program won't use the standard main interface, which is tailored for command-line applications that receive arguments. Instead of the standard main, we'll use the entry attribute from the `riscv-rt` crate to define a custom entry point. In this program we have named the entry point "main", but any other name could have been used. The entry point function must have the signature `fn() -> !`; this type indicates that the function never returns – which means that the program never terminates.
```

Comment:
This is called a Diverging Function in Rust land, not sure if we want to mention that or not.

Reply:
Wouldn't hurt to mention the term I guess - the whole paragraph was just copied from one of the other Rust embedded books as far as I remember